libgalois/include/galois/graphs/FileGraph.h

```diff
@@ -810,7 +782,7 @@ void makeSymmetric(FileGraph& in_graph, FileGraph& out) {
 
   g.setNumNodes(in_graph.size());
   g.setNumEdges(numEdges);
-  g.setSizeofEdgeData(EdgeData::has_value ? sizeof(edge_value_type) : 0);
+  g.setSizeofEdgeData(SizeOf<EdgeTy>::value);
```

Comment:
Instead of doing size, you can check if the EdgeTy is void and set it to 0.

Reply:
Here's what I tried:

- `std::is_void<EdgeTy>::value ? 0 : sizeof(EdgeTy)`
This is a *runtime* conditioning and won't compile - the compiler will still try to parse `sizeof(void)` even though 0 should be returned when EdgeTy is void.

- `sizeof(std::conditional<std::is_void<EdgeTy>::value, void*, EdgeTy>)`
In the current implementation, `EdgeData::value_type`, which is identical to `LargeArray<EdgeTy>::value_type` and also aliased as `edge_data_type`, maps void as `void*` and non-void as itself; however `sizeof(void*)` is not 0.

My design logic:

1. `LargeArray<T>::has_value` maps void to false and non-void to true by template specialization `LargeArray<void>`, so template spetializaton for void should make sense.

2. `LargeArray<T>` has a nested struct `size_of`, which for void is
```
  struct size_of {
    const static size_t value = 0;
  };
```
and for non-void is
```
  struct size_of {
    const static size_t value = sizeof(T);
  };
```
This `EdgeData::size_of::value` is also used in current implementation if you went through the comparison carefully. Therefore, generalizing it outside should make sense unless there're cases where "sizeof(void)" should not be forced to 0.

3. `sizeof` is parsed at compile time and my design reflects this.

I'd like to have your comments and thoughts on the way it should be :) thanks!

Reply:
Thanks for the detailed explanation. I'm still not convinced that defining a custom sizeof operator is the only solution available.

> This is a _runtime_ conditioning and won't compile

Yes, have you tried using std::enable_if instead? Here's one suggestion. Let setSizeofEdgeData be a template function with no arguments. If the type is void, then it's value is set to 0; otherwise, it is set of sizeof(type). This can be achieved using two enable_if functions.

Reply:
In principle, I preserve the APIs as much as possible when making changes to implementations without discussing the design of the APIs.

Now that a different API is considered, I agree with your scheme and have a further thought: Instead of making the old `setSizeofEdgeData(<size>)` a new `setSizeofEdgeData<EdgeTy>()` with ugly trailing parathesis, does it make sense to combine this to `setNumEdges(<numEdges>)` as `setNumEdges<EdgeTy>(<numEdges>)`, since `setNumNodes` `setNumEdges` `setSizeofEdgeData` are always called together?

Reply:
Yes, `setNumEdges<EdgeTy>(<numEdges>)` is a better design.